src/Cms/ModelLock.php

```diff
@@ -0,0 +1,319 @@
+<?php
+
+namespace Kirby\Cms;
+
+use Kirby\Data\Data;
+use Kirby\Toolkit\F;
+
+use Throwable;
+use Kirby\Exception\PermissionException;
+
+/**
+ * Takes care of content lock and unlock information
+ *
+ * @package   Kirby Cms
+ * @author    Nico Hoffmann <nico@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier GmbH
+ * @license   https://getkirby.com/license
+ */
+class ModelLock
+{
+
+    /**
+     * Lock data
+     *
+     * @var array
+     */
+    protected $data;
+
+    /**
+     * The model
+     *
+     * @var ModelWithContent
+     */
+    protected $model;
+
+    /**
+     * @param ModelWithContent $model
+     */
+    public function __construct($model)
+    {
+        $this->model = $model;
+        $this->data  = $this->read();
+    }
+
+    /**
+     * Sets lock with the current user
+     *
+     * @return bool
+     */
+    public function create(): bool
+    {
+        $id = $this->id();
+
+        // Check if model is already locked by another user
+        if (
+            $this->has('lock') &&
+            $this->data[$id]['lock']['user'] !== $this->user()->id()
+        ) {
+            throw new PermissionException($id . ' is already locked');
+        }
+
+        $this->data[$id]['lock'] = [
+            'user' => $this->user()->id(),
+            'time' => time()
+        ];
+
+        return $this->write();
+    }
+
+    /**
+     * Returns path to lock file
+     *
+     * @return string
+     */
+    protected function file(): string
+    {
+        return $this->model()->contentFileDirectory() . '/.lock';
+    }
+
+    /**
+     * Returns array  with `locked` flag and,
+     * if needed, `user`, `email`, `time`, `canUnlock`
+     *
+     * @return array
+     */
+    public function get(): array
+    {
+        $data = $this->data[$this->id()] ?? [];
+        $data = $data['lock'] ?? [];
+
+        if (
+            empty($data) === false &&
+            $data['user'] !== $this->user()->id() &&
+            $user = $this->kirby()->user($data['user'])
+        ) {
+            return [
+                'locked'    => true,
+                'user'      => $user->id(),
+                'email'     => $user->email(),
+                'time'      => $time = intval($data['time']),
+                'canUnlock' => $time + $this->kirby()->option('lock.duration', 60 * 2) <= time()
+            ];
+        }
+
+        return [
+            'locked' => false
+        ];
+    }
+
+    /**
+     * Checks if key in data array exists
+     *
+     * @return bool
+     */
+    protected function has(string $type): bool
+    {
+        return isset($this->data[$this->id()]) === true &&
+        isset($this->data[$this->id()][$type]) === true;
+    }
+
+    /**
+     * Returns prepended model id
+     *
+     * @return string
+     */
+    protected function id(): string
+    {
+        return '/' . $this->model()->id();
```

Comment:
What was the reason for the leading slash?

Reply:
Makes it work for Site

Reply:
Ah, of course.